### ISSUE TITLE
CSV files with BOM don't correctly strip BOM on import

### ISSUE THREAD

**Issue by eximius313** [OpenRefine/OpenRefine#1241]
Steps to reproduce:
1) Import CSV file [testUNID.zip](https://github.com/OpenRefine/OpenRefine/files/1266702/testUNID.zip)
2) Set encoding to UTF-8
3) Open project
4) Click "Edit column -> Add column based on this column"
5) In "New column name" type "Foo"
6) In "Expression" type cells["UNID"].value

Value is not loaded because the column name is prepended with a Byte Order Mark (BOM).

**Comment by ettorerizza** [comment]
I have no problem with this file and this GREL formula.

![screenshot-127 0 0 1-3333-2017-08-31-12-27-46](https://user-images.githubusercontent.com/6286410/29919009-e3eb03a0-8e47-11e7-9a19-6ff1885484a8.jpeg)

**Comment by eximius313** [comment]
wrong file - sorry. Try this one: [testUNID.zip](https://github.com/OpenRefine/OpenRefine/files/1266755/testUNID.zip)

I believe it's because UTF-8 with BOM

**Comment by ettorerizza** [comment]
> I believe it's because UTF-8 with BOM

Yep, there is an invisble character juste before the word UNID, so cells['UNID'].value doesn't work. If you rename the column UNID, it works.

**Comment by eximius313** [comment]
But it's rather a bug.
User doesn't have to know the details of encoding and should not be forced to change the name

**Comment by thadguidry** [comment]
@eximius313 Did OpenRefine guess the correct encoding in the importer options at the bottom of the Preview as UTF-8 with Bom ?  or just UTF-8 ?

**Comment by ostephens** [comment]
Just investigating this for the moment

I tested the supplied file with and without UTF-8 set on import, and found the same result both ways. Potentially related to a known issue in Java which will not be fixed http://bugs.java.com/view_bug.do?bug_id=4508058

Potentially could use http://commons.apache.org/proper/commons-io/javadocs/api-release/index.html to detect BOM and handle appropriately

**Comment by thadguidry** [comment]
I don't think there is much we can do here other than escaping a column name.
Perhaps something like

```\u00A0UNID```

Would that help you catch something next time @eximius313 ?

**Comment by ostephens** [comment]
Maybe two aspects to this:

1) Displaying any special characters in Column names - this might overlap with #1286 ?
2) Making sure BOM is never treated as part of the first column name on import?

I've not looked at the code, but feels like the latter could be a relatively easy think to check for on import and fix?

**Comment by thadguidry** [comment]
@ostephens 
RE 1 - No it won't overlap with #1286 since that will be a CSS style applied, not a value replacement.  For this issue, I would rather do a value replacement by just escaping hidden characters on a created column name as in my example and then the user can see clearly and can rename it.

RE 2 No, that can cause a few other problems.  We try really hard to give the users all their data that they import, even on Column names.  They might want to clean up their source generation or whatever, so let's be a good citizen and just inform them visually, casually.  So let's just give them their data...but slightly tweak for hidden characters in Column names.  Once #1286 lands then they also will see hidden characters in cell values as well.  But for column names, let's just escape text, instead of CSS styling.

**Comment by eximius313** [comment]
@thadguidry how about `Making sure BOM is never treated as part of the first column name on import` + proper warning that it happened?
I don't see any scenario why I would want to have invalid characters in my column names

**Comment by thadguidry** [comment]
"invalid" means different things to different people and machines.  That's why.

Maybe someone before thought it was nice to have a black heart because they LOVE something about a column name.  http://graphemica.com/%E2%9D%A4

But that black heart will actually display.  The issue is when things are hidden and do not display.

I am suggest that its a bad practice to hide things from users...when they are hidden...you don't know or cannot see if it has an impact.  So this is more about showing users those things that are hidden.

The best way to do that in any programming language is un-hide those characters...and that's always done with _escaping_ into displayable character sequences.  Like /u00A0 or whatever.

**Comment by iomicifikko** [comment]
Furthermore, the "UTF-8 with BOM" option is missing inside the "Select Encoding" window in the "Configuring Parsing Options" section (see screenshot).

I think this is the first thing to add.

![image](https://user-images.githubusercontent.com/61201406/83657996-ea83d180-a5c1-11ea-9c52-81e5f5bf4062.png)

**Comment by thadguidry** [comment]
@iomicifikko Sounds great!  Can you add all that to a [brand new issue](https://github.com/OpenRefine/OpenRefine/issues/new/choose) for us? Thanks!

**Comment by tfmorris** [comment]
It seems highly unlikely that users want a BOM in their data, so I'm not sure we need an option to control this. The comment from 2017 by @ostephens is right on the money. https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-335403657

We should use [commons.apache.org.BOMInputStream](https://commons.apache.org/proper/commons-io/javadocs/api-release/org/apache/commons/io/input/BOMInputStream.html) to strip it for the UTF cases (and take it into account for encoding guessing, if we don't do so already).

**Comment by iomicifikko** [comment]
> It seems highly unlikely that users want a BOM in their data, so I'm not sure we need an option to control this. The comment from 2017 by @ostephens is right on the money. [#1241 (comment)](https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-335403657)

True
 
> We should use [commons.apache.org.BOMInputStream](https://commons.apache.org/proper/commons-io/javadocs/api-release/org/apache/commons/io/input/BOMInputStream.html) to strip it for the UTF cases (and take it into account for encoding guessing, if we don't do so already).

Encoding guessing is great but I don't know if there is any chance of wrong choice by the software.
I can actually switch in OpenRefine between, for instance, **x-UTF-16LE-BOM** and **UTF-16LE** in the parsing configuration. Same thing for **UTF-32**. Actually there is not similar option for **UTF-8**.

@thadguidry : thanks for the link but maybe it is better to wait a little longer, the discussion is still evolving.

**Comment by jquartel** [comment]
I've just run into this problem myself, and one thing I'd like to add to the discussion is that even though it's a bit crazy to put a BOM on a UTF-8 file (there can be no byte ordering when the words are only one byte - no doubt that's why Java doesn't support it as an encoding), I have found that when one exports a spreadsheet as CSV from Microsoft Excel it adds this silly BOM to the file. So I think this will come up a lot for those of us who receive files made by people using Excel. Here's an example.
[Excel_CSV_Test.csv](https://github.com/OpenRefine/OpenRefine/files/7659818/Excel_CSV_Test.csv)
.

**Comment by RolfBly** [comment]
We also ran into this problem. There are quite a few database systems that export UTF-8 with BOM by default, _because it disambiguates what you're dealing with_. 

Attached is a sample (shortened) export from a leading collection management system. The screenshot shows you what OR makes of it. Remove the BOM & everythings fine. But it's not at all obvious you have to do this. So people try to work around it by loading as line based text files. ending up using | as field separator and trimming leading and trailing quotes afterwards. It's ugly and it shouldn't be necessary. 

I simply can't understand why this has not been taken seriously for over five years. I do a lot of data wrangling (that OR can't do) using Python, and I have no problem reading like this: 

`with open(infile, encoding='utf-8-sig', mode='r')`

Please comment. 

![Schermafbeelding 2022-10-17 174133](https://user-images.githubusercontent.com/198277/196222057-7aef4619-3051-4633-8212-8d00c614610d.png)

[BOMbug.csv](https://github.com/OpenRefine/OpenRefine/files/9802216/BOMbug.csv)

**Comment by thadguidry** [comment]
@RolfBly Thanks for your patience, first of all.  The BOM (Byte order mark) that you are referring to is actually the UTF-8 file signature (commonly also called a "BOM") which identifies the encoding format rather than the byte order of the document.  Your example in Python actually knows this as you see `utf-8-sig` meaning signature or file signature.  UTF-8 is a linear sequence of bytes and not a sequence of 2-byte or 4-byte units where the byte order is important.  In your provided csv file, the actual encoded BOM is `EF BB BF`. [[1]](https://learn.microsoft.com/en-us/globalization/encoding/byte-order-mark)

![image](https://user-images.githubusercontent.com/986438/196313487-fc86dd97-2adc-4e74-a5d4-0eed9d9e4f71.png)


With that out of the way, we understand you are looking (as others are in this issue) for OpenRefine to handle the `EF BB BF` at the beginning of the file and give some options to remove it when it is detected.
Would you prefer:
1. a warning that it was detected during preview and give you the option to remove it?
2. no warning, just remove it.
3. no warning, just a checkbox option to let me toggle "remove UTF-8 BOM (file signature)" and see the preview differences.
4. a warning, and a checkbox option to let me toggle "remove UTF-8 BOM (file signature)" and see the preview differences.

**Comment by RolfBly** [comment]
Hi @thadguidry, please see [Wikipedia's lemma byte order mark and UTF-8](https://en.wikipedia.org/wiki/Byte_order_mark/#UTF-8). Perhaps you see opportunity to improve it. 

In my experience, hex ef bb bf at the beginning of a file is one of the very few marks that prevent lots of software packages from mutilating diacritical characters beyond all recognition. Even Excel started handling diacritics properly only with Office 2016. Things have been a mess since the beginning of computers. So how bad is it when you are called Dàniël but people keep spelling your name as Dániel or Daniel? Let's ask Dàniël. (I bet you a beer his mom's Hungarian. But that aside.)

"No exception should go unnoticed!" says the Zen of Python. Will that work though, in this case? For a normal person? Who is not a programmer, but is charge of the proper names of the persons and institutions in a collection? I think not. So Just remove it and show a checkbox 

> "UTF-8 encoding mark removed (deselect to keep it)". 

Not-so-technical people won't even notice that checkbox, or won't understand what it means at all. Those who do understand have the choice. 

Shouldn't you ask others too? 

Cheers,
Rolf

**Comment by ostephens** [comment]
I think we can (and should, after 5 years!) just make a decision and get this to a point where someone could (if they chose) actively work on it.

I think the suggestion is a good one and would propose:

* Add optional removal of BOM from UTF-8 on *SV importer  (using https://commons.apache.org/proper/commons-io/javadocs/api-release/org/apache/commons/io/input/BOMInputStream.html)
* Add check box to *SV importer screen "Remove BOM from UTF-8 encoded files", checked by default (perhaps with the option of changing the default through a preference https://docs.openrefine.org/manual/running#preferences)
* Ensure that changing the option updates the preview
* Option should only apply if UTF-8 is set as encoding - so a stretch goal would be to only have the option active in the UI if encoding set as UTF-8

If we can agree on this as a way forward then we have an actionable piece of development

**Comment by jquartel** [comment]
I would suggest we just add a new character encoding called, say "UTF-8, with BOM" (this happens to be exactly what BBEdit calls it). This would be auto-detected and indicated in the Character encoding field in the 'Parse data as' section of the import view. And when this character encoding is selected, the BOM is stripped. But one could still go to the Character encoding list and select "UTF-8" for the old behaviour (keeping the BOM).

<img width="1094" alt="Import suggestion" src="https://user-images.githubusercontent.com/36742596/197173147-93781187-945a-4a8c-ac5a-a21797677066.png">

Similarly, if one would like to create a file with the BOM, one could just select "UTF-8, with BOM" from the list of encodings in Custom tabular exporter > Download > Character encoding.

<img width="816" alt="Export suggestion" src="https://user-images.githubusercontent.com/36742596/197173160-2a4b76fd-22f1-41b5-8151-dda4ce998716.png">

This functionality happens to mimic exactly what BBEdit (a common text editor) does. I think it feels logical to detect and remove this kind of file metadata when it's found since it doesn't really constitute (or is highly unlikely to constitute) any part of the data that you're trying to refine.

Note that while looking at BBEdit for inspiration, I noticed it also included "UTF-16, with BOM" and "UTF-16, no BOM" in its encoding set, so perhaps we should consider UTF-16 in the same way.

**Comment by RolfBly** [comment]
What @jquartel describes is similar to that other common text editor, Notepad++. So yes, good idea!

**Comment by wetneb** [comment]
I'd rate "good second issue" perhaps - happy to be contradicted.

**Comment by tfmorris** [comment]
> Please comment.

Sure, happy to.

> We also ran into this problem. There are quite a few database systems that export UTF-8 with BOM by default

@RolfBly  Can you give some examples of these "quite a few database systems," please? 

> Attached is a sample (shortened) export from a leading collection management system. 

Please replace the weasel words with the name of the system so that people know what you're talking about.

> I simply can't understand why this has not been taken seriously for over five years. 

I'm not sure why you think it wasn't taken seriously. A proposed solution was investigated and outlined in [2017](https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-335403657) and reconfirmed in [2020](https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-638300820). I'm sure the development team would be happy to review a contribution from you (or anyone else) implementing it. It's a collaborative volunteer community, after all.

> I do a lot of data wrangling (that OR can't do) using Python, and I have no problem reading like this:
> 
> `with open(infile, encoding='utf-8-sig', mode='r')`

Awesome. Sounds like you have multiple workarounds for the issue. Not only can you preprocess the file before opening it in OpenRefine, but you can also use Python.

Since we're rehashing the previous discussion, I'll also point out that the list of encodings comes from Java [as mentioned before](https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-638365023) and the encoding guessing is done by a [third-party library](https://www-archive.mozilla.org/projects/intl/chardet.html), although we used to use a [different library](https://unicode-org.github.io/icu/userguide/conversion/detection.html), I think.

Perhaps it would help put this in perspective to note that encoding guessing [didn't work at all for 8 1/2 years](https://github.com/OpenRefine/OpenRefine/issues/486#issuecomment-638397305).

**Comment by RolfBly** [comment]
Hi @tfmorris,

> quite a few database systems that export UTF-8 with BOM by default

My work is in collection management systems, that's collections of museums, libraries and archives. My main expertise is with products called Adlib and Axiell Collections. I do data repair and data conversion regularly. So far, I've done data conversions from TMS, Pica/WinIBW, Worldcat, and a few applications based on Microsoft Access, Filemaker Pro and Cardbox. 

I can't give you a list of which versions of which package do what with regards to UTF-8 signature. My experience is that if any export has it, there are usually no diacritics falling over. Some versions of the above mentioned have it, others, especially the older ones, don't. 

If it doesn't, I've learned to first go looking for e.g. ë and ö to find out how things may be encoded. Oddball mixups, e.g. little pieces of text copy-pasted from e.g. WordPerfect with just a few characters deviating, may occur. 

In OR this is hidden, but you'll see it when OR does not replace all strings that look identical, but under the water aren't. I sometimes copy paste it into a hex editor to know for sure. I can tell you that none of my customers never go that far. 

> Please replace the weasel words

See above. I  intentionally did not name the product (1) because this is a general issue that keeps popping up regularly across the board, and (2) out of discretion. I'm not here to blame anyone. 

> I'm not sure why you think it wasn't taken seriously

Taking issues seriously usually leads to solutions. No solution for a long time? Must be getting not high enough priority. Which can be frustrating to some. I think you're making the very same point [here](https://github.com/OpenRefine/OpenRefine/issues/486#issuecomment-638397305), even.  

> I'm sure the development team would be happy to review a contribution from you

I'm self-employed. Meaning I would have to do that in my spare time. Of which I have very little. 
Also, I know next to nothing about Java (only "it's like C" and "versioning is a mess"). I get by with Python, but I'm by no means what one could call a seasoned coder. 

I live in the canyon between the rock with people who make it and sell it, and the rock on the other side with people who use it. I've already told some of the latter how to work around this particular issue, for now. If you think the workaround should be in some manual or wiki, I'd be happy to write it up. 

> encoding guessing does not work

Please CMIIW, `EF BB BF` as first three bytes means the rest is UTF-8 encoded, right? No encoding guessing needed? Isn't that the whole point of having those three bytes at all? Then I must be missing something very essential here. 

Ta, 
Rolf

**Comment by jquartel** [comment]
In relation to this statement:

> Since we're rehashing the previous discussion, I'll also point out that the list of encodings comes from Java https://github.com/OpenRefine/OpenRefine/issues/1241#issuecomment-638365023 and the encoding guessing is done by a [third-party library](https://www-archive.mozilla.org/projects/intl/chardet.html), although we used to use a [different library](https://unicode-org.github.io/icu/userguide/conversion/detection.html), I think.

If we think it's too messy to add a 'with BOM' encoder to the list of library-defined encoders, then I think it would be just as effective to add a 'with BOM' check box to the right of the character encoding text box. It would be greyed out and empty whenever the encoding is not one that can use a BOM and, for the import case, it would automatically get checked when a BOM is detected (and the checked state effectively indicating that the BOM is removed). On the export page, the check box would effectively indicate whether you want to include the BOM.

**Comment by jquartel** [comment]
I know this is a really old issue, but I'm still finding new sources of CSV data where the BOM is included. I've just received such a file exported from Notion (https://www.notion.so/product), for instance, and we already know that exports from Microsoft Excel include the BOM. I have so many of these now that I've written a script that strips the BOMs just so that I can bring them into OpenRefine, but it would be so great if Refine could handle this in some way. OpenRefine is such a brilliant tool that it's central to the way that I and my team work, so I'd love to see it progress to being a little more user friendly.
